File: Incidents/TableView.m
////
// TableView.m
// Incidents
////
// See the file COPYRIGHT for copyright information.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
////

#import <objc/message.h>
#import "TableView.h"



@implementation TableView


- (void) keyDown:(NSEvent *)event
{
    id delegate = self.delegate;

    if (delegate && self.selectedRow != -1) {
        unichar key = [[event charactersIgnoringModifiers] characterAtIndex:0];
        if (key == NSDeleteCharacter) {
            if ([delegate respondsToSelector:@selector(deleteFromTableView:)]) {
                [delegate deleteFromTableView:self];
                return;
            }
        }
        else if (
            key == NSCarriageReturnCharacter ||
            key == NSEnterCharacter          ||
            key == NSNewlineCharacter
        ) {
            if ([delegate respondsToSelector:@selector(openFromTableView:)]) {
                [delegate openFromTableView:self];
                return;
            }
        }
    }

    [super keyDown:event];
}

@end

File: Incidents/TableView.swift
//
//  TableView.swift
//  Incidents
//
//  © 2015 Burning Man and its contributors. All rights reserved.
//  See the file COPYRIGHT.md for terms.
//

import Cocoa



class TableView: NSTableView {
    
    override func keyDown(theEvent: NSEvent) {
        guard let delegate = self.delegate() as? TableViewDelegate else {
            logError("Table view has no delegate: \(self)")
            return
        }
        
        if selectedRow == -1 { return }
        
        // FIXME: Seem janky that I have to cast to NSString, etc here
        
        // We're looking for exactly one character
        guard let characterString: NSString = theEvent.charactersIgnoringModifiers else { return }
        guard characterString.length == 1 else { return }
        let key = characterString.characterAtIndex(0)
        
        switch key {
            case unichar(NSDeleteCharacter):
                delegate.deleteFromTableView?(self)
            case unichar(NSCarriageReturnCharacter), unichar(NSEnterCharacter):
                delegate.openFromTableView?(self)
            default:
                break
        }
        
        super.keyDown(theEvent)
    }
    
}



@objc
protocol TableViewDelegate: NSTableViewDelegate {

    optional func deleteFromTableView(tableView: TableView)
    optional func openFromTableView(tableView: TableView)

}
